nncf/experimental/torch/nas/bootstrapNAS/search/search.py

```diff
@@ -0,0 +1,386 @@
+        mutation_eta = search_config.get('mutation_eta', 3.0)
+        acc_delta = search_config.get('acc_delta', 1)
+        ref_acc = search_config.get('ref_acc', 100)
+
+        return cls(num_evals, num_constraints, population,
+                 seed, crossover_prob, crossover_eta,
+                 mutation_prob, mutation_eta, acc_delta, ref_acc)
+
+
+class BaseSearchAlgorithm:
+    """
+    Base class for search algorithms. It contains the evaluators used by search approches.
+    """
+    def __init__(self):
+        self._use_default_evaluators = True
+        self._evaluators = None
+        self._bad_requests = []
+
+    @abstractmethod
+    def run(self, validate_fn: Callable, val_loader: DataLoader, checkpoint_save_dir: str,
+            evaluators: Optional[List[Evaluator]] = None, ref_acc: Optional[float] = 100,
+            tensorboard_writer: Optional[SummaryWriter] = None) -> Tuple[ElasticityController,
+                                                                         SubnetConfig, Tuple[float, ...]]:
+        pass
+
+
+class SearchAlgorithm(BaseSearchAlgorithm):
+    def __init__(self,
+                 model: NNCFNetwork,
+                 elasticity_ctrl: ElasticityController,
+                 nncf_config: NNCFConfig,
+                 verbose=True):
+        super(SearchAlgorithm, self).__init__()
+        self._model = model
+        self._elasticity_ctrl = elasticity_ctrl
+        self._elasticity_ctrl.multi_elasticity_handler.activate_maximum_subnet()
+        search_config = nncf_config.get('bootstrapNAS', {}).get('search', {})
+        self._num_obj = None
+        self.search_params = SearchParams.from_dict(search_config)
+        self._log_dir = nncf_config.get("log_dir", ".")
+        self._tb = None
+        self._verbose = verbose
+        self._top1_accuracy_validation_fn = None
+        self._val_loader = None
+        evo_algo = search_config['algorithm']
+        if evo_algo == 'NSGA2':
+            self._algorithm = NSGA2(pop_size=self.search_params._population,
+                                    sampling=get_sampling("int_lhs"),
+                                    crossover=get_crossover("int_sbx", prob=self.search_params._crossover_prob,
+                                                            eta=self.search_params._crossover_eta),
+                                    mutation=get_mutation("int_pm", prob=self.search_params._mutation_prob, eta=self.search_params._mutation_eta),
+                                    eliminate_duplicates=True,
+                                    save_history=True,
+                                    )
+        else:
+            raise NotImplementedError(f"Evolutionary Search Algorithm {evo_algo} not implemented")
+        self._num_vars = 0
+        self._vars_lower = 0
+        self._vars_upper = []
+
+        for dim in elasticity_ctrl.multi_elasticity_handler.get_available_elasticity_dims():
+            if dim == ElasticityDim.KERNEL:
```

Comment:
There shouldn't be multiple if condition. Instead, introduce ElasticSearchHandler base class, subclass ElasticWidthSearchHandler, ElasticKernelSearchHandler and so on.
```python
handlers = dict() #  type: Dict[ElasticityDim, ElasticSearchHandler]
for dim in elasticity_ctrl.multi_elasticity_handler.get_available_elasticity_dims():
    self._num_vars += handlers[dim].get_num_vars()
    self._vars_upper += handlers[dim].get_num_vars()
```

Reply:
Refactored and simplified how to get this information. One issue that still remains is that each dimension produces a different data structure to describe its search space.